File: cpp/Source/gamit/message/MessageManager.cpp
#include "MessageManager.h"

using namespace gamit;

CMessageManager * CMessageManager::_inst(nullptr);

CMessageManager::CMessageManager()
{

}

CMessageManager::~CMessageManager()
{

}


CMessageManager * CMessageManager::instance()
{
	if (nullptr == _inst)
	{
		_inst = new CMessageManager();
	}

	return _inst;
}

void CMessageManager::registHandler(int command, const CMessageHandlerBasePtr & handler)
{
	_mapMessageHandler[command] = handler;
}

void CMessageManager::__onMessage(CSerializer & __is)
{
	MessageBlockPtr msgBlock = new MessageBlock(__is);

	int command = msgBlock->_command;
	auto found = _mapMessageHandler.find(command);

	if (found != _mapMessageHandler.end())
	{
		found->second->onMessage(msgBlock);
	}
}

File: android/src/rmi/MessageManager.java
package rmi;

import java.util.HashMap;
import java.util.Map;

public class MessageManager {
	private Map<Integer, MessageHandler> _mapHandler;
	
	static MessageManager _inst;
	static {
		_inst = new MessageManager();
	}
	
	private MessageManager()
	{
		_mapHandler = new HashMap<Integer, MessageHandler>();
	}
	
	public static MessageManager instance()
	{
		return _inst;
	}
	
	public void registMessageHandler(int command, MessageHandler handler)
	{
		_mapHandler.put(new Integer(command), handler);
	}
	
	public void onMessage(Serializer __is)
	{
		MessageBlock msgBlock = new MessageBlock(__is);
		int command = msgBlock.getCommand();
		Integer key = new Integer(command);
		
		MessageHandler handler = _mapHandler.get(key);
		if (handler != null)
		{
			handler.onMessage(msgBlock);
		}
	}
}
